multicore_runtime/wsgi_config.py

```diff
@@ -0,0 +1,208 @@
+  to import it or failure to find `webapp_add_wsgi_middleware` is not an
+  error.
+
+  Returns:
+    The appengine_config.webapp_add_wsgi_middleware function or None.
+  """
+  try:
+    import appengine_config  # pylint: disable=g-import-not-at-top
+    try:
+      return appengine_config.webapp_add_wsgi_middleware
+    except AttributeError:
+      return None
+  except ImportError:
+    return None
+
+
+def static_app_for_handler(handler):
+  """Returns a WSGI app that serves static files as directed by the handler.
+
+  Args:
+    handler: An individual handler from appinfo_external.handlers
+      (appinfo.URLMap)
+
+  Returns:
+    A static file-serving WSGI app closed over the handler information.
+  """
+  regex = handler.url
+  files = handler.static_files
+  upload = handler.upload
+  if not files:
+    if handler.static_dir:
+      # If static_files is not set, convert static_dir to static_files and also
+      # modify the url regex accordingly. See the appinfo.URLMap docstring for
+      # more information.
+      regex = static_dir_url(handler)
+      files = handler.static_dir + r'/\1'
+      upload = handler.static_dir + '/.*'
+    else:
+      # Neither static_files nor static_dir is set; log an error and return.
+      logging.error('No script, static_files or static_dir found for %s',
+                    handler)
+      return None
+  return static_app_for_regex_and_files(regex, files, upload,
+                                        mime_type=handler.mime_type)
+
+
+def static_dir_url(handler):
+  """Converts a static_dir regex into a static_files regex if needed.
+
+  See the appinfo.URLMap docstring for more information.
+
+  Args:
+    handler: A handler (appinfo.URLMap)
+
+  Returns:
+    A modified url regex
+  """
+  if not handler.script and not handler.static_files and handler.static_dir:
+    return handler.url + '/(.*)'
+  else:
+    return handler.url
+
+
+def load_user_scripts_into_handlers(handlers):
+  """Preloads user scripts, wrapped in env_config middleware if present.
+
+  Args:
+    handlers: appinfo.handlers data as provided by get_module_config()
+
+  Returns:
+    A list of tuples suitable for configuring the dispatcher() app,
+    where the tuples are (url, script, app):
+      - url: The url pattern which matches this handler.
+      - script: The script to serve for this handler, as a string, or None.
+      - app: The fully loaded app corresponding to the script.
+  """
+  # `if x.login == appinfo.LOGIN_OPTIONAL` disables loading handlers
+  # that require login or admin status entirely. This is a temporary
+  # measure until handling of login-required handlers is implemented
+  # securely.
+  loaded_handlers = [
+      (x.url if x.script or x.static_files else static_dir_url(x),
+       x.script,
+       app_for_script(x.script) if x.script else static_app_for_handler(x))
+      for x in handlers if x.login == appinfo.LOGIN_OPTIONAL]
+  logging.info('Parsed handlers: %s',
```

Comment:
Although `%s` works, a list's str() representation will match its repr() and I think `%r` might be slightly more accurate here since what you're printing is essentially the Python representation of the data structure.